### PR TITLE
Reduce boost regex dependencies

### PR DESCRIPTION
**Pull request by camolezi**
#### PR description:
Replaced boost regex for c++11 stl regex, they should have similar behavior.

#### PR validation:
Passed on basic runTheMatrix test.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

@vgvassilev @davidlange6

### CODE REVIEW

**Comment by alja** [comment]
+1

File: FWCore/Framework/doc/EventSelector.cc-with-cerr-traces [hunk 1]
```diff
@@ -13,7 +13,7 @@
 #include "FWCore/Utilities/interface/Algorithms.h"
 
 #include "boost/algorithm/string.hpp"
-#include "boost/regex.hpp"
+#include <regex>
```
Comment (makortel) [inline FWCore/Framework/doc/EventSelector.cc-with-cerr-traces:16]: This file is obsolete. I suggest we let this change go in and then I remove the file (among others). Unless this PR needs to be modified for other reasons, in which case I'd ask to leave this file untouched (but only in that case).

**Comment by makortel** [comment]
+core

File: HLTrigger/Timer/plugins/FastTimerServiceClient.cc [hunk 2]
```diff
@@ -3,7 +3,7 @@
 #include <cstring>
 
 // boost headers
-#include <boost/regex.hpp>
+#include <regex>
```
Comment (fwyzard) [inline HLTrigger/Timer/plugins/FastTimerServiceClient.cc:6]: Could you move this header to the "C++ headers" section, and remove the now empty "boost headers" section ?

File: HLTrigger/Timer/plugins/ThroughputServiceClient.cc [hunk 3]
```diff
@@ -3,7 +3,7 @@
 #include <cstring>
 
 // boost headers
-#include <boost/regex.hpp>
+#include <regex>
```
Comment (fwyzard) [inline HLTrigger/Timer/plugins/ThroughputServiceClient.cc:6]: Could you move this header to the "C++ headers" section, and remove the now empty "boost headers" section ?

File: PhysicsTools/FWLite/src/TH1Store.cc [hunk 4]
```diff
@@ -4,7 +4,7 @@
 #include <iomanip>
 #include <cassert>
```
Comment (fwyzard) [inline PhysicsTools/FWLite/src/TH1Store.cc:6]:
```suggestion
```

File: CommonTools/Utils/src/TFileDirectory.cc [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 #include "CommonTools/Utils/interface/TFileDirectory.h"
 
-#include "boost/regex.hpp"
+#include <regex>
```
Comment (fwyzard) [inline CommonTools/Utils/src/TFileDirectory.cc:6]: Could you move this header with the other standard headers above ?

File: CondCore/CondDB/interface/Utils.h [hunk 6]
```diff
@@ -13,7 +13,7 @@
 #include <pwd.h>
 #include <climits>
 //
```
Comment (fwyzard) [inline CondCore/CondDB/interface/Utils.h:15]:
```suggestion
```

File: EventFilter/L1TRawToDigi/interface/MP7FileReader.h [hunk 7]
```diff
@@ -16,7 +16,7 @@
 #include <vector>
 
 // Boost Headers
-#include <boost/regex.hpp>
+#include <regex>
```
Comment (fwyzard) [inline EventFilter/L1TRawToDigi/interface/MP7FileReader.h:19]: Could you move this header with the other standard headers above ?

File: EventFilter/L1TRawToDigi/src/MP7FileReader.cc [hunk 8]
```diff
@@ -4,7 +4,7 @@
 #include "EventFilter/L1TRawToDigi/interface/MP7FileReader.h"
 
 #include <boost/algorithm/string.hpp>
-#include <boost/regex.hpp>
+#include <regex>
```
Comment (fwyzard) [inline EventFilter/L1TRawToDigi/src/MP7FileReader.cc:7]: Could you move this header with the other standard headers below ?

**Review by fwyzard — CHANGES_REQUESTED** [review]

**Comment by camolezi** [comment]
I don't think we should go on with this replacement anymore, unless performance is not a concern.

I've made some benchmarks on std::regex vs boost::regex.
They show that std::regex is at least 10x slower than boost::regex.

https://gist.github.com/camolezi/87283aaf7cdeb584ba294588a5d77abd
https://gist.github.com/camolezi/ae83ce140a5ade6e9c4aee158f9f95e2

**Comment by makortel** [comment]
@camolezi Interesting (and quick internet searches support that). I'd be curious if adding `-O2` would make any difference?

**Comment by camolezi** [comment]
> @camolezi Interesting (and quick internet searches support that). I'd be curious if adding `-O2` would make any difference?

wow, It made a big difference. The performance difference is much more reasonable with -O2 flag.
In the regex_match benchmark, with -O2 flag.

                                    cpu-time      iterations
boost_regex                2805 ns       250338
std_regex                    2214 ns       312415

**Comment by camolezi** [comment]
> > @camolezi Interesting (and quick internet searches support that). I'd be curious if adding `-O2` would make any difference?
> 
> wow, It made a big difference. The performance difference is much more reasonable with -O2 flag.
> In the regex_match benchmark, with -O2 flag.
> 
> boost_regex: 2805ns - iterations: 250338
> std_regex: 2214ns - iterations: 312415
> 
> edit: it seems that std_regex is faster with -O2 flag. I benchmarked again a couple of times and this result seems right.

Okay, my regex_match benchmarks do show that std_regex is comparable to boost regex.
But even with optimizations, in my other benchmark (regex_seach), still suggest a huge difference in performance.

Based on this:
https://www.boost.org/doc/libs/1_73_0/libs/regex/doc/html/boost_regex/background/performance.html

It seems like in some specific cases std::regex does have a comparable performance to boost::regex. But in general, it seems like std::regex is significantly slower.

**Comment by fwyzard** [comment]
Interesting - thanks a lot for the investigation.

I am, of course, in favour of staying with the better-performing option :-)

**Pull request closed without merging**